src/codegen/compile.ml

```diff
@@ -10657,15 +10671,13 @@ and compile_prim_invocation (env : E.t) ae p es at =
     G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^
     (* Not using Tagged.load_field since it is not a proper pointer to the array start *)
     Heap.load_field (Arr.header_size env) (* loads the element at the byte offset *)
-  | GetPastArrayOffset spacing, [e] ->
-    let shift =
-      match spacing with
-      | ElementSize
-      | One -> BigNum.from_word30 env in    (* make it a compact bignum *) (* TBR: from_word30? *)
+  | GetLastArrayOffset, [e] ->
+    assert (BitTagged.can_tag_const Type.Int (Int64.of_int32 (Int32.sub (Arr.max_array_size env) 1l)));
     SR.Vanilla,
     compile_exp_vanilla env ae e ^^ (* array *)
     Arr.len env ^^
-    shift
+    compile_sub_const 1l ^^
```

Comment:
Hmm, does this add an extra subtraction in every iteration?

Reply:
No, it's only called once before loop entry (see desugar.ml)

Reply:
No, it's only called once before loop entry (see desugar.ml)